File: src/zros_cpp/subscriber_manager.cc
//
// Created by pairs on 7/28/18.
//

#include <zros/subscriber_manager.h>
#include <zros/singleton.h>
#include <sspdlog/sspdlog.h>
#include "impl/subscriber_impl.h"

namespace zros {
    // todo
    SubscriberManager::SubscriberManager() {
        impl_ = std::make_shared<SubscribersImpl>();
        service_discovery_ = SingleTon<ServiceDiscovery>::Instance();
        service_discovery_->set_register_publisher_cb([this](const zros_rpc::PublisherInfo* publisherInfo, zros_rpc::Status* status){
            impl_->registerPublisher(publisherInfo, status);
        });
        service_discovery_->set_unregister_publisher_cb([this](const zros_rpc::PublisherInfo* publisherInfo, zros_rpc::Status* status){
           impl_->unregisterPublisher(publisherInfo, status);
        });
    }

    bool SubscriberManager::registerSubscriber(const std::shared_ptr<ISubscriber> subscriber) {
        // 1. register to master first
        bool ok = service_discovery_->addSubscriber(subscriber);
        if (ok) {
            // 2. register to memory
            SSPD_LOG_INFO << "register subscriber " << subscriber->get_topic() << " to master success";
            return impl_->registerSubscriber(subscriber);
        }
        SSPD_LOG_WARNING << "register subscriber " << subscriber->get_topic() << " to master failed";
    }

    bool SubscriberManager::unregisterSubscriber(const std::string &topic) {
        return impl_->unregisterSubscriber(topic);
    }

}

File: src/zros_python/subscriber_manager.py
# -*- coding: utf-8 -*-
""" The subscriber management

data: 2018-08-02
author: pairs
"""

import zros_python.impl.subscriber_impl as impl
import zros_python.zros as zros
import logging

logger = logging.getLogger(__name__)


class SubscriberManager(object):
    """

    """
    def __init__(self):
        self._impl = impl.SubscribersImpl()
        self._gsd = zros.get_service_discovery()
        self._gsd.set_register_publisher_cb(self._impl.register_publisher)
        self._gsd.set_unregister_publisher_cb(self._impl.unregister_publisher)

    def register_subscriber(self, subscriber):
        # 1. register to master first
        ok = self._gsd.add_subscriber(subscriber)
        if ok:
            logger.info(u'register subscriber %s to master success', subscriber.get_topic())
            return self._impl.register_subscriber(subscriber)
        logger.warning(u'register subscriber %s to master failed', subscriber.get_topic())
        return False

    def unregister_subscriber(self, topic):
        return self._impl.unregister_subscriber(topic)
